### PR TITLE
handle accounts without access to GCP

### LINKED ISSUE #1401: NullPointerException in ProjectRepository.convertToGcpProjects
image

I believe this happened when trying to deploy while logged into an account that does not have GCP access. I forget the exact name, but Google Apps for Your Domain account or some such?

```
java.lang.NullPointerException
	at com.google.cloud.tools.eclipse.projectselector.ProjectRepository.convertToGcpProjects(ProjectRepository.java:93)
	at com.google.cloud.tools.eclipse.projectselector.ProjectRepository.getProjects(ProjectRepository.java:54)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.StandardDeployPreferencesPanel.retrieveProjects(StandardDeployPreferencesPanel.java:365)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.StandardDeployPreferencesPanel.access$3(StandardDeployPreferencesPanel.java:360)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.StandardDeployPreferencesPanel$2.run(StandardDeployPreferencesPanel.java:281)
	at com.google.cloud.tools.eclipse.login.ui.AccountSelector.fireSelectionListeners(AccountSelector.java:147)
	at com.google.cloud.tools.eclipse.login.ui.AccountSelector.selectAccount(AccountSelector.java:133)
	at com.google.cloud.tools.eclipse.login.ui.AccountSelectorObservableValue.doSetValue(AccountSelectorObservableValue.java:59)
	at org.eclipse.core.databinding.observable.value.AbstractObservableValue.setValue(AbstractObservableValue.java:66)
	at org.eclipse.core.databinding.UpdateValueStrategy.doSet(UpdateValueStrategy.java:469)
	at org.eclipse.core.databinding.ValueBinding$3$1.run(ValueBinding.java:206)
	at org.eclipse.core.databinding.observable.Realm$1.run(Realm.java:149)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.databinding.observable.Realm.safeRun(Realm.java:153)
	at org.eclipse.core.databinding.observable.Realm.exec(Realm.java:171)
	at org.eclipse.core.databinding.ValueBinding$3.run(ValueBinding.java:196)
	at org.eclipse.core.databinding.observable.Realm$1.run(Realm.java:149)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.databinding.observable.Realm.safeRun(Realm.java:153)
	at org.eclipse.core.databinding.observable.Realm.exec(Realm.java:171)
	at org.eclipse.core.databinding.ValueBinding.doUpdate(ValueBinding.java:158)
	at org.eclipse.core.databinding.ValueBinding.updateModelToTarget(ValueBinding.java:123)
	at org.eclipse.core.databinding.DataBindingContext.updateTargets(DataBindingContext.java:464)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.DeployPreferencesDialog.createContents(DeployPreferencesDialog.java:82)
	at org.eclipse.jface.window.Window.create(Window.java:426)
	at org.eclipse.jface.dialogs.Dialog.create(Dialog.java:1095)
	at org.eclipse.jface.window.Window.open(Window.java:783)
	at com.google.cloud.tools.eclipse.appengine.deploy.ui.standard.StandardDeployCommandHandler.execute(StandardDeployCommandHandler.java:91)
	at org.eclipse.ui.internal.handlers.HandlerProxy.execute(HandlerProxy.java:295)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:55)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:282)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:264)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:132)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:152)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:494)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:488)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:210)
	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.executeItem(HandledContributionItem.java:433)
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem.handleWidgetSelection(AbstractContributionItem.java:454)
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem$3.handleEvent(AbstractContributionItem.java:482)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4256)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1501)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1509)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1313)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4080)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3706)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1121)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1022)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:150)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:693)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:610)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:138)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:673)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:610)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1519)
```

### PR DESCRIPTION
**Pull request by elharo**
fix #1401 @akerekes @chanseokoh

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java [hunk 1]
```diff
@@ -107,7 +109,7 @@ protected Control createDialogArea(final Composite parent) {
           @Override
           public int getMessageType(ValidationStatusProvider statusProvider) {
             int type = super.getMessageType(statusProvider);
-            setValid(type != IMessageProvider.ERROR);
+            setValid(type != IMessageProvider.ERROR && content.hasSelection());
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java]: @briandealwis Isn't this by-passing or abusing databinding? I feel if the project selector does not have any selection, it should give the ERROR state with the message "select a project", which I believe was the behavior prior to the new project selector.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java [hunk 2]
```diff
@@ -176,9 +178,9 @@ public boolean isHelpAvailable() {
   }
 
   private void setValid(boolean isValid) {
-    Button okButton = getButton(IDialogConstants.OK_ID);
-    if (okButton != null) {
-      okButton.setEnabled(isValid);
+    Button deployButton = getButton(IDialogConstants.OK_ID);
+    if (deployButton != null) {
+      deployButton.setEnabled(isValid && content.hasSelection());
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java]: Ditto.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java (hunk 1 above)
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java]: Just like this:

deploy to app engine standard _002

Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java]: I think disabling the button is a regression and a separate issue: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1408

**Comment by elharo** [comment]
The problem is some accounts don't have any projects to select from. We should probably think about how to better indicate that.

**Comment by elharo** [comment]
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.739 sec <<< FAILURE! - in com.google.cloud.tools.eclipse.appengine.localserver.server.LocalAppEnginePublishOperationTest
testPublishingSubmodules(com.google.cloud.tools.eclipse.appengine.localserver.server.LocalAppEnginePublishOperationTest)  Time elapsed: 2.733 sec  <<< FAILURE!
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertNotNull(Assert.java:712)
	at org.junit.Assert.assertNotNull(Assert.java:722)
	at com.google.cloud.tools.eclipse.appengine.localserver.server.LocalAppEnginePublishOperationTest.setUp(LocalAppEnginePublishOperationTest.java:66)

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java [hunk 3]
```diff
@@ -489,4 +489,8 @@ public void setFont(Font font) {
     expandableComposite.setFont(font);
     FontUtil.convertFontToBold(expandableComposite);
   }
+
+  boolean hasSelection() {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: hasSelectedProject?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java (hunk 2 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java]: Agreed. This doesn't seem right — the caller into setValid() should be doing the appropriate tests.

**Comment by elharo** [comment]
Looks like #1378 disconnected the ProjectIdInputValidator. The class still exists but is no longer referenced.

**Comment by chanseokoh** [comment]
Yeah, I think reconnecting it (probably aftet remaining it because it is no longer an input) should do it.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java (hunk 1 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java]: I agree.  The `hasSelection()` test should be moved into the `StandardDeployPreferencesPanel`.  I think the easiest is to add a _ValidationStatusProvider_ just as `setupAccountEmailDataBinding()` does.

Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java]: #1410 will fix this.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java (hunk 2 above)
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java]: #1410 will fix this.

**Comment by elharo** [comment]
Using data binding is proving far more complicated than what I have here. Doubtless possible but by no means the easiest option, and will likely cost us an extra day at least. :-(

**Pull request merged**